### PR TITLE
refactor: use observer pattern to play scene

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Using `isPlaying` to start the scene playing is a bit counter-intuitive: you'd expect it to track the state of the scene, not cause it to start.

My first attempt at fixing this https://github.com/freeCodeCamp/freeCodeCamp/pull/57295 meant we suddenly had to deal with `null` scenes, which wasn't pretty. 

This PR makes use of the observer pattern so that we can keep all the scene state inside scene and allow any component to start the scene playing, via a `.notify` call. This means we don't have to handle null `scene` data, except insofar as we don't create the scene component. Also, we can share the `SceneSubject` with anything that needs to start the scene - at the moment, hotkeys and an on-click listener in Scene.

Also, it should make it easier to implement pausing (ref: https://github.com/freeCodeCamp/freeCodeCamp/issues/55293) since it's easy to extend the subject to notify based on different events.

### CODE REVIEW

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
This is not working for me very reliably. I suspect it is because I have a slow connection and machine using the small Gitpod instance.

To reproduce:
1) Navigate to the first English A2 dialogue
2) (maybe) before the page/scene fully loads, press `ctrl + space` (playScene hotkey)
3) (maybe) press the hotkey again
4) (usually) see error

<details>
  <summary>Error</summary>

image


We do this elsewhere, to prevent such an error:

```ts
if (audioRef.current) {
      // This is to prevent a stupid UI bug (all Shaun's fault)
      audioRef.current.load();
      audioRef.current
        .play()
        .then(() => {
          if (audioRef.current) {
            audioRef.current.pause();
          }
        })
        .catch(console.error);
    }
```

</details>

**Comment by ojeytonwilliams** [comment]
I couldn't recreate the error, @ShaunSHamilton, but I did find another bug: it was possible to start multiple scenes running at once. I fixed that simply by ignoring multiple play requests. Not the most elegant solution, but it seems to work.

**Comment by ShaunSHamilton** [comment]
> Where do we do that?

Exam Environment app.

**Comment by ojeytonwilliams** [comment]
@ShaunSHamilton no rush, but I've not been able to reproduce the error. If it still happens for you, let me know.

**Comment by moT01** [comment]
I was able to reproduce this on firefox, not chrome or safari.

**Comment by ojeytonwilliams** [comment]
Thanks, Tom. Same steps as Shaun?

**Comment by moT01** [comment]
yup. Load the first dialogue and press the play hotkey before it's loaded.

File: client/src/templates/Challenges/components/scene/scene.tsx [hunk 1]
```diff
@@ -84,142 +86,165 @@ export function Scene({
...
-      window.requestAnimationFrame(maybeStopAudio);
-    }
-  }
-
-  const playScene = () => {
-    setShowDialogue(true);
+  const pause = () => {
+    // Until the play() promise resolves, we can't pause the audio
+    if (canPauseRef.current) audioRef.current.pause();
+    canPauseRef.current = false;
```
Comment (ShaunSHamilton) [inline client/src/templates/Challenges/components/scene/scene.tsx:120]: I am not sure if this has any impact, but do we not want this:

```suggestion
    if (canPauseRef.current) {
      audioRef.current.pause();
      canPauseRef.current = false;
    }
```

**Review by ShaunSHamilton — APPROVED** [review]
👍 

Any issues I am producing, can also be produced on `.org` and `.dev`. So, unrelated to this PR.

File: client/src/templates/Challenges/components/scene/scene.tsx (hunk 1 above)
Comment (ojeytonwilliams) [reply client/src/templates/Challenges/components/scene/scene.tsx:120]: They're equivalent, so we can use whichever. If `current` is `true`, then both versions set it to `false`. If `current` is `false`, then the unnested version sets it to `false` again, which doesn't matter.

Comment (ojeytonwilliams) [reply client/src/templates/Challenges/components/scene/scene.tsx:120]: I don't mind which version we go with, honestly.

**Review by naomi-lgbt — APPROVED** [review]

**Pull request merged**